src/classlibnative/bcltype/bignum.cpp

```diff
@@ -0,0 +1,597 @@
+
+    UINT64 carry = 0;
+    while (pCurrent != pEnd)
+    {
+        UINT64 product = (UINT64)(*pCurrent) * (UINT64)value + carry;
+        carry = product >> 32;
+        *pResultCurrent = (UINT32)(product & 0xFFFFFFFF);
+
+        ++pResultCurrent;
+        ++pCurrent;
+    }
+
+    if (lhs.m_len < BIGSIZE && result.m_blocks[lhs.m_len] != 0)
+    {
+        result.m_len = lhs.m_len + 1;
+    }
+    else
+    {
+        result.m_len = lhs.m_len;
+    }
+}
+
+void BigNum::Multiply(const BigNum& lhs, const BigNum& rhs, BigNum& result)
+{
+    const BigNum* pLarge = NULL;
+    const BigNum* pSmall = NULL;
+    if (lhs.m_len < rhs.m_len)
+    {
+        pSmall = &lhs;
+        pLarge = &rhs;
+    }
+    else
+    {
+        pSmall = &rhs;
+        pLarge = &lhs;
+    }
+
+    UINT8 maxResultLength = pSmall->m_len + pLarge->m_len;
+
+    // Zero out result internal blocks.
+    memset(result.m_blocks, 0, sizeof(UINT32) * BIGSIZE);
+
+    const UINT32* pLargeBegin = pLarge->m_blocks;
+    const UINT32* pLargeEnd = pLarge->m_blocks + pLarge->m_len;
+
+    UINT32* pResultStart = result.m_blocks;
+    const UINT32* pSmallCurrent = pSmall->m_blocks;
+    const UINT32* pSmallEnd = pSmallCurrent + pSmall->m_len;
+
+    while (pSmallCurrent != pSmallEnd)
+    {
+        // Multiply each block of large BigNum.
+        if (*pSmallCurrent != 0)
+        {
+            const UINT32* pLargeCurrent = pLargeBegin;
+            UINT32* pResultCurrent = pResultStart;
+            UINT64 carry = 0;
+
+            do
+            {
+                UINT64 product = (UINT64)(*pResultCurrent) + (UINT64)(*pSmallCurrent) * (UINT64)(*pLargeCurrent) + carry;
+                carry = product >> 32;
+                *pResultCurrent = (UINT32)(product & 0xFFFFFFFF);
+
+                ++pResultCurrent;
+                ++pLargeCurrent;
+            } while (pLargeCurrent != pLargeEnd);
+
+            *pResultCurrent = (UINT32)(carry & 0xFFFFFFFF);
+        }
+
+        ++pSmallCurrent;
+        ++pResultStart;
+    }
+
+    if (maxResultLength > 0 && result.m_blocks[maxResultLength - 1] == 0)
+    {
+        result.m_len = maxResultLength - 1;
+    }
+    else
+    {
+        result.m_len = maxResultLength;
+    }
+}
+
+bool BigNum::IsZero()
+{
+    if (m_len == 0)
+    {
+        return true;
+    }
+
+    for (UINT8 i = 0; i < m_len; ++i)
+    {
+        if (m_blocks[i] != 0)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void BigNum::SetUInt32(UINT32 value)
+{
+    m_len = 1;
+    m_blocks[0] = value;
+}
+
+void BigNum::SetUInt64(UINT64 value)
+{
+    m_len = 0;
+    m_blocks[0] = (UINT32)(value & 0xFFFFFFFF);
+    m_len++;
+
+    UINT32 highBits = (UINT32)(value >> 32);
+    if (highBits != 0)
+    {
+        m_blocks[1] = highBits;
+        m_len++;
+    }
+}
+
+void BigNum::ExtendBlock(UINT32 newBlock)
+{
+    m_blocks[m_len] = newBlock;
+    ++m_len;
+}
+
+UINT32 BigNum::LogBase2(UINT32 val)
```

Comment:
How about using `__builtin_clz` for FEATURE_PAL and `_BitScanReverse` / `_BitScanReverse64` for Windows? My guess is that it would be the fastest way to compute the logarithm.